### ISSUE TITLE
Build a reproducible KoRA implementation and benchmark compositional transfer against matched PEFT baselines

### ISSUE THREAD

**Issue by OnePunchMonk** [OnePunchMonk/KoRA#2]
## Proposal

Build KoRA into a reproducible compositional-adapter research package, then test one falsifiable hypothesis: **does communication between adapter signals improve held-out transfer over independent adapters and an equally sized side network, at comparable parameter and compute budgets?**

This should establish the core PEFT result before expanding to KoRA-Route in #1. The architecture variants below are proposals, not validated improvements. This issue is based on static inspection; no training experiments were run for it.

## 1. Establish what the current implementation actually does

Reviewed revision: [`830ce29`](https://github.com/OnePunchMonk/KoRA/tree/830ce2910f093de588cbfa6ce24c09b293362dc7). Evidence: [README](https://github.com/OnePunchMonk/KoRA/blob/830ce2910f093de588cbfa6ce24c09b293362dc7/README.md) and [vision notebook](https://github.com/OnePunchMonk/KoRA/blob/830ce2910f093de588cbfa6ce24c09b293362dc7/src/kora-vision.ipynb), especially `KoRAViT.forward`, `run_experiment`, and `get_tiny_imagenet_loaders`.

Several discrepancies should be resolved before interpreting the headline results:

- **Architecture mismatch:** the README describes Q/K/V composition into a gated V update. The notebook captures detached inputs from attention QKV/projection and MLP layers, pools adapter outputs over tokens, composes them sequentially, and adds a delta to the final CLS representation. It uses fixed `alpha`, not the described zero-initialized learnable gate. Preserve this implementation as a named **side-network baseline**, and separately implement the attention-local proposal.
- **Optimizer omission:** `final_projection` is created inside the first forward, after parameter counting and optimizer construction. Move it into `__init__`; assert every intended trainable parameter is in the optimizer and that it changes after an update.
- **Invalid transfer evaluation:** the notebook evaluates a CIFAR-100 head against Tiny ImageNet labels before replacing the head. These label spaces are unrelated; labels above 99 can also cause cross-entropy to fail. Remove this as a transfer metric.
- **Uncontrolled few-shot sampling:** `list(range(1000))` is a dataset-order prefix, not a stratified sample. Replace it with seeded, class-balanced sampling (five examples per class for a 1,000-example Tiny ImageNet training set), saved sample IDs, class histograms, and an explicit shared class mapping.
- **CKA interpretation:** the current KoRA side network does not modify intermediate backbone activations. CKA measured on backbone blocks cannot establish that its composition branch learned better cross-layer representations. Measure the adapted representations as well; similarity alone is not proof of useful transfer.

Treat the README's 97–98% Tiny ImageNet results as **unverified pending reproducible split manifests, checkpoints, and per-example predictions**. These observations do not establish how the historical numbers were produced. Preserve the historical results and explain their evidence status instead of silently replacing them.

## 2. Build a small, testable implementation

Suggested layout: `src/kora/{layers,vit}.py`, `configs/`, `scripts/{train,evaluate,benchmark}.py`, `tests/`, and `results/<run_id>/`.

Start with one pinned `timm` ViT checkpoint and explicit module wrappers. Register all modules before creating the optimizer; avoid mutable capture hooks in the new implementation. Implement adapter-only save/load with backbone revision, target-module names, dimensions, rank, and scaling in the checkpoint metadata.

For a first **attention-local KoRA** variant, retain token resolution. For row-vector token features `x`, use:

```text
z_q = x A_q^T; z_k = x A_k^T; z_v = x A_v^T        # each [..., r]
u = GELU([z_q; z_k; z_v] C_1^T) C_2^T            # [..., r]
q = x W_q^T + (alpha/r) z_q B_q^T
k = x W_k^T + (alpha/r) z_k B_k^T
v = x W_v^T + (alpha/r) z_v B_v^T + (beta/r) u B_c^T
```

Compose in rank space to avoid concatenating three full-width deltas. With composer hidden width `h`, the bias-free composer adds `4rh` parameters plus `d_v r` for `B_c` per block. Report the actual totals, including heads and any normalization parameters.

Use random down projections/composer weights and zero-initialized up projections, with fixed nonzero initial scales. This preserves the initial base-model output while allowing up-projection gradients. Do not combine a zero gate with an identically zero branch and accidentally block all learning. Test expected staged gradient flow across multiple optimizer steps.

Required correctness checks: initial equivalence to the frozen model; zero/off composition recovering the LoRA path; frozen backbone weights remaining unchanged; optimizer coverage; save/load prediction equivalence; batch/token shapes; and finite gradients. Verify fused-QKV slicing explicitly.

## 3. Improve only through controlled ablations

| Comparison | Question |
| --- | --- |
| Independent LoRA vs linear rank-space mixing vs nonlinear mixing | Does nonlinearity add value beyond sharing or extra rank? |
| KoRA vs parameter-matched MLP side network using the same tapped features | Is the gain communication or just added capacity? |
| Full mixing vs block-diagonal/no-cross-branch mixing | Does cross-adapter communication itself matter? |
| Fixed scale vs learnable gate; with/without normalization | Does stabilization improve optimization? |
| Final-CLS side network vs token-preserving local updates | Does placement explain classification or dense-task behavior? |
| Different ranks, composer widths, and adapted-layer counts | Where is the quality/cost frontier? |

Run these sequentially after the baseline works, rather than a large Cartesian sweep. A spline/basis-function composer can follow only if nonlinear mixing wins. The present MLP is KA-inspired; the representation theorem does not establish a generalization advantage or make it a KAN implementation. See [KAN](https://arxiv.org/abs/2404.19756) for a distinct architecture reference.

## 4. Benchmark protocol

**Stage A — correctness and affordable screening:** tiny synthetic overfit test, then CIFAR-100 source adaptation and class-balanced Tiny ImageNet transfer. Use the same backbone checkpoint, augmentations, precision, effective batch size, source budget, and seed-specific examples across methods. Document overlap between backbone pretraining data and evaluation datasets.

Separate three target protocols:

1. Frozen original backbone + newly trained target head.
2. Source-trained adapters frozen + newly trained target head: tests representation transfer.
3. Source-trained adapters + target head adapted on target training data: tests few-shot adaptation.

Use identical target-head initialization per seed. Explicitly freeze non-head parameters in protocol 2. Include target adaptation from the original checkpoint without source training to test whether CIFAR preadaptation helps at all. Use a training-derived validation split for tuning; reserve official Tiny ImageNet validation for final evaluation and verify image/label mappings and split disjointness.

**Stage B — independent confirmation:** preselect a small [VTAB](https://google-research.github.io/task_adaptation/) panel spanning natural, specialized, and structured tasks; use its published training/validation protocol. Freeze the method/config selection before final test evaluation, then expand to the full suite if worthwhile. State clearly whether results use standard backbone-to-task adaptation or additional source preadaptation.

**Baselines:** frozen backbone + linear head, standard [LoRA](https://arxiv.org/abs/2106.09685), [DoRA](https://arxiv.org/abs/2402.09353), the corrected notebook side network, and the parameter-matched MLP control. Full fine-tuning is a reference when affordable. Align insertion sites and heads; report both equal-rank and approximately equal-total-trainable-parameter comparisons. Give each method the same declared hyperparameter-search budget.

**Quality:** top-1 accuracy, macro-F1, per-class accuracy, source performance, and target improvement relative to the original-backbone control. Report five final seeds, mean/std, paired per-seed differences, and confidence intervals. Use three seeds only for exploratory screening. Keep failed/negative runs. CKA is a diagnostic, using fixed examples and explicitly named representations, not a success criterion; see the [CKA paper](https://arxiv.org/abs/1905.00414).

**Cost:** total/trainable parameters (head separately), peak training memory, examples/sec, elapsed training time, and inference latency at fixed hardware, resolution, batch size, and precision. Warm up and synchronize accelerator timings. Report both matched training-step and matched wall-clock comparisons. Nonlinear composition generally cannot merge into a fixed linear weight delta; compare against both merged and unmerged LoRA and report that deployment cost.

**Later dense prediction:** revisit NYU Depth V2 using token/spatial adapters and the same decoder for every method. Pin split, crop, depth units, valid-pixel masks, and metric implementation; report RMSE, AbsRel, and delta thresholds. Do this after classification/transfer evidence is reliable.

## 5. Deliverables and decision gate

- [ ] Corrected baseline package with pinned dependencies, documented commands, and correctness tests.
- [ ] Dataset manifests with hashes, seeds, class mappings, sample counts, and no train/evaluation overlap.
- [ ] One-command paired LoRA/KoRA run producing configs, commit/checkpoint IDs, environment/hardware metadata, logs, predictions, metrics, and checkpoints.
- [ ] Communication/capacity ablations and a quality-versus-parameters/time/memory report with uncertainty.
- [ ] README tables regenerated from run artifacts, with historical/unverified and reproduced results distinguished.

**Proposed gate:** advance the method only if a predeclared target aggregate shows a positive paired improvement with a 95% interval above zero against the tuned parameter-matched LoRA and MLP controls, within a declared compute budget, and the result survives independent task confirmation. Publish source-task regressions and per-task losses alongside the aggregate. An inconclusive or negative outcome is useful evidence: simplify or stop the variant instead of expanding its claims. Revisit #1 after this core experiment is settled.

**Comment by OnePunchMonk** [comment]
Pushed initial theory-grounded implementation in [`d5fadde`](https://github.com/OnePunchMonk/KoRA/commit/d5fadde): [`src/kora/multiplicative.py`](../blob/d5fadde/src/kora/multiplicative.py) implements `f(x)=alpha BAx + beta U((Px)⊙(Qx))`; [`tests/test_multiplicative.py`](../blob/d5fadde/tests/test_multiplicative.py) checks zero-initial output, gradient flow, and exact scalar-product representation; [`docs/formal_analysis.md`](../blob/d5fadde/docs/formal_analysis.md) records the conditional proof and counterexamples. Local validation: `PYTHONPATH=src python3 -m pytest -q` → 2 passed.
